### DISCUSSION TITLE
Error registering tasks to build new objects

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#18884] [Q&A General] [answered]
Dear all,
I am intending to create a new object "ConstantVectorPostprocessor_UO" which is based conceptually on "ConstantVectorPostprocessor" but does not inherit any of their member functions/variables as such. The former one is now being developed to read data from user objects rather than from the input file itself. The block of commands following is relevant to this new object.

<pre> <code>
[VectorPostprocessors]
  [from_main_vpp_UO]
    type = ConstantVectorPostprocessor_UO
    stresstens_UO = stresstens
    vector_names = 'a b c'
  []
[]
</pre></code> 

The following section of code initialises the paramaters and the member variables.

```
#include "ConstantVectorPostprocessor_UO.h"
#include "InputParameters.h"

registerMooseObject("MooseApp", ConstantVectorPostprocessor_UO);

//registerMooseObjectTask("add_vector_postprocessor", ConstantVectorPostprocessor_UO, false);

defineLegacyParams(ConstantVectorPostprocessor_UO);

InputParameters
ConstantVectorPostprocessor_UO::validParams()
{
//  InputParameters params = emptyInputParameters();
  InputParameters params = MooseObject::validParams();
  params.addClassDescription(
      "Populate constant VectorPostprocessorValue directly from an user object.");
  params.addParam<std::vector<std::string>>("vector_names",
                                            "Names of the column vectors in this object");
  params.addRequiredParam<UserObjectName>(
      "stresstens_UO", "Name of user object that defines the stress tensor.");

  params.registerBase("ConstantVectorPostprocessor_UO");
  return params;
}

ConstantVectorPostprocessor_UO::ConstantVectorPostprocessor_UO(const InputParameters & parameters)
   : MooseObject(parameters),
   Restartable(this, "ConstantVectorPostprocessor_UO"),
   _vector_names(getParam<std::vector<std::string>>("vector_names"))
{
  _vector_names = getParam<std::vector<std::string>>("vector_names");
  unsigned int nvec = _vector_names.size();
}
```

Executing this test gives the following run time error.

<pre><code>
*** ERROR ***
Task add_vector_postprocessor is not registered to build ConstantVectorPostprocessor_UO derived objects
</pre></code>
I am pretty sure the above task is not properly resgistered to the new bject that is currently under development. I tried to set this right in Moose.C by setting the boolean to true in the following line of code.
<pre><code>
registerMooseObjectTask("add_vector_postprocessor",     VectorPostprocessor,    true);
</pre></code>
However this does not fix the issue even with a 'clobberall' build. Happy to welcome any suggestions in this regard.
Kind regards,
Arun

**Comment by GiudGiud** [comment, marked as answer]
Hello

The issue that the inheritance chain for your new postprocessor does not include a postprocessor class.
So if you change MooseObject to GeneralPostprocessor then your object will actually be a postprocessor to the system. 

Note that while you may not want to re-use any of the member functions of the ConstantPP, some of the ones above the chains are pretty important

**Reply by abarun22** [reply]
Hi Guillaume,
Inclusion of 'GeneralPostprocessor' seems to induce compilation issues deep in to the system files. The following error has been encountered.
<pre><code>
(base) abalasub@freia013> make -j10
Using HIT from /home/abalasub/Working/Projects/moose/framework/contrib/hit
Creating Unity /home/abalasub/Working/Projects/moose/framework/build/unity_src/vectorpostprocessors_Unity.C
Compiling C++ (in opt mode) /home/abalasub/Working/Projects/moose/framework/build/unity_src/vectorpostprocessors_Unity.C...
In file included from /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/x86_64-pc-linux-gnu/bits/c++allocator.h:33:0,
                 from /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/allocator.h:46,
                 from /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/string:41,
                 from /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/locale_classes.h:40,
                 from /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/ios_base.h:41,
                 from /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/ios:42,
                 from /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/istream:38,
                 from /usr/local/depot/gcc-7.3.0/include/c++/7.3.0/fstream:38,
                 from /home/abalasub/Working/Projects/moose/framework/src/vectorpostprocessors/CSVReader.C:11,
                 from /home/abalasub/Working/Projects/moose/framework/build/unity_src/vectorpostprocessors_Unity.C:2:
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = exampleVectorPostprocessor; _Args = {const InputParameters&}; _Tp = exampleVectorPostprocessor]’:
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/alloc_traits.h:475:4:   required from ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = exampleVectorPostprocessor; _Args = {const InputParameters&}; _Tp = exampleVectorPostprocessor; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<exampleVectorPostprocessor>]’
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr_base.h:526:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {const InputParameters&}; _Tp = exampleVectorPostprocessor; _Alloc = std::allocator<exampleVectorPostprocessor>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2]’
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr_base.h:637:4:   required from ‘std::__shared_count<_Lp>::__shared_count(std::_Sp_make_shared_tag, _Tp*, const _Alloc&, _Args&& ...) [with _Tp = exampleVectorPostprocessor; _Alloc = std::allocator<exampleVectorPostprocessor>; _Args = {const InputParameters&}; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2]’
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr_base.h:1295:35:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<exampleVectorPostprocessor>; _Args = {const InputParameters&}; _Tp = exampleVectorPostprocessor; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2]’
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr.h:344:64:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<exampleVectorPostprocessor>; _Args = {const InputParameters&}; _Tp = exampleVectorPostprocessor]’
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr.h:690:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = exampleVectorPostprocessor; _Alloc = std::allocator<exampleVectorPostprocessor>; _Args = {const InputParameters&}]’
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/bits/shared_ptr.h:706:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = exampleVectorPostprocessor; _Args = {const InputParameters&}]’
/home/abalasub/Working/Projects/moose/framework/build/header_symlinks/Registry.h:146:29:   required from ‘std::shared_ptr<MooseObject> moose::internal::buildObj(const InputParameters&) [with T = exampleVectorPostprocessor]’
/home/abalasub/Working/Projects/moose/framework/build/header_symlinks/Registry.h:207:21:   required from ‘static char Registry::add(const RegistryEntry&) [with T = exampleVectorPostprocessor]’
/home/abalasub/Working/Projects/moose/framework/src/vectorpostprocessors/exampleVectorPostprocessor.C:12:1:   required from here
/usr/local/depot/gcc-7.3.0/include/c++/7.3.0/ext/new_allocator.h:136:4: error: invalid new-expression of abstract class type ‘exampleVectorPostprocessor’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/abalasub/Working/Projects/moose/framework/src/vectorpostprocessors/exampleVectorPostprocessor.C:10:0,
                 from /home/abalasub/Working/Projects/moose/framework/build/unity_src/vectorpostprocessors_Unity.C:42:
/home/abalasub/Working/Projects/moose/framework/build/header_symlinks/exampleVectorPostprocessor.h:21:7: note:   because the following virtual functions are pure within ‘exampleVectorPostprocessor’:
 class exampleVectorPostprocessor : public GeneralPostprocessor
       ^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/FEProblemBase.h:28:0,
                 from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/DistributionInterface.h:14,
                 from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/UserObject.h:13,
                 from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/GeneralUserObject.h:13,
                 from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/GeneralVectorPostprocessor.h:14,
                 from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/CSVReader.h:13,
                 from /home/abalasub/Working/Projects/moose/framework/src/vectorpostprocessors/CSVReader.C:14,
                 from /home/abalasub/Working/Projects/moose/framework/build/unity_src/vectorpostprocessors_Unity.C:2:
/home/abalasub/Working/Projects/moose/framework/build/header_symlinks/Postprocessor.h:36:30: note:      virtual PostprocessorValue Postprocessor::getValue()
   virtual PostprocessorValue getValue() = 0;
                              ^~~~~~~~
In file included from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/GeneralUserObject.h:13:0,
                 from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/GeneralVectorPostprocessor.h:14,
                 from /home/abalasub/Working/Projects/moose/framework/build/header_symlinks/CSVReader.h:13,
                 from /home/abalasub/Working/Projects/moose/framework/src/vectorpostprocessors/CSVReader.C:14,
                 from /home/abalasub/Working/Projects/moose/framework/build/unity_src/vectorpostprocessors_Unity.C:2:
/home/abalasub/Working/Projects/moose/framework/build/header_symlinks/UserObject.h:67:16: note:         virtual void UserObject::execute()
   virtual void execute() = 0;
                ^~~~~~~
/home/abalasub/Working/Projects/moose/framework/build/header_symlinks/UserObject.h:72:16: note:         virtual void UserObject::initialize()
   virtual void initialize() = 0;
                ^~~~~~~~~~
make: *** [/home/abalasub/Working/Projects/moose/framework/build/unity_src/vectorpostprocessors_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1
</pre></code>
Line 12 in exampleVectorPostprocessor.C (renamed my class "ConstantVectorPostprocessor_UO" just for an example and to debug the issue) corresponds to 

<pre><code>
registerMooseObject("MooseApp", exampleVectorPostprocessor);
</pre></code>

Commenting this line compiles the code, but falters when it searches for the registration of this new object. Another alternative is usage of MooseObject as the parent class for the new class, in which case it compiles without any problems, but asks me to register the task `add_vector_postprocessor` to the newly derived class at run time. I tried to check this very deep in to the `Action` classes to see if there is any missing link, and tried some experiments but eventually could not fix it.  I suspect there are some problems in the base level functions, especially when it comes to registering objects under `VectorPostprocessors`, which might need to be corrected.
Welcome your suggesions in this regard.

**Reply by GiudGiud** [reply]
Look at the header for GeneralPostprocessor and make sure all the pure virtual functions are implemented.
Some of them you can just implement an empty routine

Guillaume

**Reply by GiudGiud** [reply]
I'm thinking there might be a derived class that is better than GeneralPostprocessor to inherit. Maybe ElementPostprocessor if you are working on elements, InterfacePP for sides on interfaces etc

That will implement a couple of these pure virtual routines (in generalPP) for you.

**Reply by abarun22** [reply]
Yes, i think it is appropriate to derive from objects that has every thing self contained. The data to be stored in VPP comes from an integration point of a particular element that i choose programmatically based on certain calculations, which in my case would happen in an user object. The new VPP that i am planning to develop, will read this user oject and populate the variables to be transferred across the apps. This is very much inline with "ConstantVectorPostprocessor" and the good thing is inheriting this object works in our case, but the input structure is quite different and so i cannot use this object moving forward. Either we must identify the right object to inherit the vector post processor features or we might want to create a new object (as can be seen from this example `SamplerBase.C` and `NodalValueSampler.C`). Both of these approaches eventually comes down to the case of finding the right VPP object to inherit from, which in our case may be none. Probably we might be required to build an object with only basic framework that can sit inside VectorPostprocessor block and continue the developments from there on.

**Comment by abarun22** [comment]
Hi Guillaume,
This is now working OK after a struggle. I allowed my new object to inherit from `GeneralVectorPostprocessor` and retained the `initialize()` and `execute` pure virtual functions. 
Kind regards,
Arun

**Reply by GiudGiud** [reply]
Great! As there is no objects that does part of the work for you then, the General... are the recommended base classes. 
ConstantVPP inherits from GeneralVPP
